include/nlohmann/json.hpp

```diff
@@ -4330,19 +4463,73 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
     @sa see @ref erase(const size_type) -- removes the element from an array at
     the given index
 
-    @since version 1.0.0
+    @since version 1.0.0; template added in version 3.10.0
     */
-    size_type erase(const typename object_t::key_type& key)
+    template < class KeyT, typename std::enable_if <
+#if defined(JSON_HAS_CPP_17)
+                   !std::is_same<typename std::decay<KeyT>::type, std::string_view>::value&&
+#endif
+                   std::is_convertible<KeyT, typename object_t::key_type>::value, int >::type    = 0 >
+    size_type erase(KeyT && key)
     {
         // this erase only works for objects
         if (JSON_HEDLEY_LIKELY(is_object()))
         {
-            return m_value.object->erase(key);
+            return m_value.object->erase(std::forward<KeyT>(key));
         }
 
         JSON_THROW(type_error::create(307, "cannot use erase() with " + std::string(type_name()), *this));
     }
 
+#if defined(JSON_HAS_CPP_17)
+    /*!
+    @brief remove element from a JSON object given a key
+
+    Removes elements from a JSON object with the key value @a key.
+
+    @param[in] key value of the elements to remove
+
+    @return Number of elements removed. If @a ObjectType is the default
+    `std::map` type, the return value will always be `0` (@a key was not
+    found) or `1` (@a key was found).
+
+    @post References and iterators to the erased elements are invalidated.
+    Other references and iterators are not affected.
+
+    @throw type_error.307 when called on a type other than JSON object;
+    example: `"cannot use erase() with null"`
+
+    @complexity `log(size()) + count(key)`
+
+    @liveexample{The example shows the effect of `erase()`.,erase__key_type}
+
+    @sa see @ref erase(IteratorType) -- removes the element at a given position
+    @sa see @ref erase(IteratorType, IteratorType) -- removes the elements in
+    the given range
+    @sa see @ref erase(const size_type) -- removes the element from an array at
+    the given index
+
+    @since version 3.10.0
+    */
+    size_type erase(const std::string_view& key)
```

Comment:
Is the performance different between `std::map::erase()` and
```cpp
if(auto const it = std::map::find(); it != std::map::end())
{
    std::map::erase(it);
}
```
?

I would assume that the non-iterator version of `std::map::erase()` is essentially using `std::map::find()` under the hood.

Reply:
I would presume that everything could be implemented in terms of `find`, so the string view API is just re-implementing std::map using only a subset of it's API.

The performance of std::erase should be equivalent to std::find in a normal implementation, but someone should double check the C++ spec.

Reply:
What do you propose?

Reply:
My thought was to implement all of the operations in terms of find (except those that are already natively templated properly in std::map itself).

That *should* allow you to shrink down to the minimum amount of additional functions, with ideally only a single templated function for each operation.

Reply:
Yeah, the same idea came across my mind too, one can remove all special casing and overloading by intentionally re-implementing std::map. Since you're doing this anyway to support std::string_view, it wouldn't hurt to remove the older code to just use this now.

Actually, it should remove special casing of std::string_view entirely, leaving the library to be plain old C++11, yet perfectly usable with C++17 std::string_views. Seems like a win-win.

Reply:
I got rid of the special cases. Looks much cleaner now.